designs/NuGetExtensibility/INuGetProjectServices.md

```diff
@@ -0,0 +1,153 @@
+# INuGetProjectServices
+
+* Status: In Review
+* Author(s): [Andy Zivkovic](https://github.com/zivkan)
+* Issue: [9577](https://github.com/NuGet/Home/issues/9577) IVsPackageInstallerServices should be async
+
+## Problem Background
+
+The existing IVsPackageInstallerServices service is synchronous, and is being called by several components on the UI thread, causing UI delays in Visual Studio. We are creating new async APIs that will allow us to unblock the UI thread, and use the Visual Studio Service Broker to work in Visual Studio CodeSpaces (and maybe Live Share).
+
+See [NuGet extensibility services](NuGetExtensibilityServices.md) for more details.
+
+## Who are the customers
+
+Visual Studio component and extension developers.
+
+## Goals
+
+* Can be called on any thread (ideally background thread, but work if called on UI thread).
+* Do not block caller (implement async properly, return incomplete task instead).
+* Works with Visual Studio Codespaces, and hopefully Live Share.
+
+## Non-Goals
+
+* This will not be a 1:1 replacement for any of NuGet's existing `IVs` interfaces.
+
+## Solution
+
+```cs
+/// <summary>Service to interact with projects in a solution</summary>
+/// <remarks>This interface should not be implemented. New methods may be added over time.</remarks>
```

Comment:
`This interface should not be implemented. New methods may be added over time.`

Isn't this what default interface methods are for? 😄

Reply:
This assembly is the contract for NuGet's Visual Studio extensibility. When we add a method like `InstallPackageAsync`, there's no way to add a default implementation that is meaningful, other than throw new not implemented exception.

This remark is a signal that anyone using this interface should not implement a wrapper by "inheriting" the interface, particularly in production code. I'd also suggest they don't implement the interface in a test project to use as a mock for their own code.